### ISSUE TITLE
Compression for IndexedSlices not supported

### ISSUE THREAD

**Issue by UditGupta10** [horovod/horovod#966]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet): Tensorflow
2. Framework version: 1.10
3. Horovod version:0.15.1
4. MPI version: MPI 3.0
5. CUDA version: NA
6. NCCL version:NA
7. Python version:3.5.6
8. OS and version: NAME="SLES"
VERSION="12-SP2"
VERSION_ID="12.2"
PRETTY_NAME="SUSE Linux Enterprise Server 12 SP2"

**Checklist:**
1. Did you search issues to find if somebody asked this question before? 
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.md)? 
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.md)? 

**Your question:**
Please ask your question here.
Q. Why is compression not supported for tf.IndexedSlices (Sparse representation)? Is it because its tricky to implement or doesn't have any advantage over non-compressed version.

**Comment by alsrgv** [comment]
@UditGupta10, I think it'd be a good addition, we just didn't get to it yet :-)

**Comment by UditGupta10** [comment]
The compression for allreduce is implemented in a very easy to understand and straightforward way, I was thinking that may be similar thing can be done with allgather as well. Could you mention a few issues/complications that may occur while implementing compression for allgather.

**Comment by alsrgv** [comment]
Allgather is even simpler conceptually, since we just need to cast the tensor, allgather it, and cast back.  There's no need for special handling in the Horovod core (e.g. like one we had to do for [CPU FP16 allreduce](https://github.com/horovod/horovod/blob/57be59474ddca30f4fdcca38269bfc1c5bc7ea4e/horovod/common/operations.cc#L772)).  We should use the same compression interface as we use for allreduce.
